projects/lesson-4/runtime/src/kitties.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+#[derive(Encode, Decode)]
+pub struct Kitty(pub [u8; 16]);
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Kitties {
+		/// Stores all the kitties, key is the kitty id / index
+		pub Kitties get(kitty): map T::KittyIndex => Option<Kitty>;
+		/// Stores the total number of kitties. i.e. the next kitty index
+		pub KittiesCount get(kitties_count): T::KittyIndex;
+
+		/// Get kitty ID by account ID and user kitty index
+		pub OwnedKitties get(owned_kitties): map (T::AccountId, T::KittyIndex) => Option<T::KittyIndex>;
+		/// Get number of kitties by account ID
+		pub OwnedKittiesCount get(owned_kitties_count): map T::AccountId => T::KittyIndex;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		/// Create a new kitty
+		pub fn create(origin) {
+			let sender = ensure_signed(origin)?;
+
+			// 作业：重构create方法，避免重复代码
+
+//			let kitty_id = Self::kitties_count();
+//			if kitty_id == T::KittyIndex::max_value() {
+//				return Err("Kitties count overflow");
+//			}
+			let kitty_id = Self::next_kitty_id()?;
+
+			// Generate a random 128bit value
+//			let payload = (<system::Module<T>>::random_seed(), &sender, <system::Module<T>>::extrinsic_index(), <system::Module<T>>::block_number());
+//			let dna = payload.using_encoded(blake2_128);
+			let dna = Self::random_value(&sender);
+
+			// Create and store kitty
+			let kitty = Kitty(dna);
+//			<Kitties<T>>::insert(kitty_id, kitty);
+//			<KittiesCount<T>>::put(kitty_id + 1.into());
+//
+//			// Store the ownership information
+//			let user_kitties_id = Self::owned_kitties_count(&sender);
+//			<OwnedKitties<T>>::insert((sender.clone(), user_kitties_id), kitty_id);
+//			<OwnedKittiesCount<T>>::insert(sender, user_kitties_id + 1.into());
+			Self::insert_kitty(sender,kitty_id,kitty)
+		}
+
+		/// Breed kitties
+		pub fn breed(origin, kitty_id_1: T::KittyIndex, kitty_id_2: T::KittyIndex) {
+			let sender = ensure_signed(origin)?;
+
+			Self::do_breed(sender, kitty_id_1, kitty_id_2)?;
+		}
+
+		//Transfer
+        fn transfer(origin, to: T::AccountId, kitty_id: T::KittyIndex) -> Result {
+            let from = ensure_signed(origin)?;
+
+			//1 -转移不能同一个人
+			ensure!(from != to, "from and to needs different");
+
+			//2 - 判断index是否有效
+			let kitty = Self::kitty(kitty_id);
+
+			ensure!(kitty.is_some(), "Invalid kitty_id");
+
+			//3 - 判断转出者是否拥有kittyindex这只猫
+			let kitty_owner = Self::owned_kitties((from.clone(),kitty_id));
+			ensure!(kitty_owner.is_some(), "Invalid kitty_id");
+
+            Self::transfer_from(from, to, kitty_id)?;
+
+            Ok(())
+        }
+	}
+}
+
+
+fn combine_dna(dna1: u8, dna2: u8, selector: u8) -> u8 {
+	// 作业：实现combine_dna
+	// 伪代码：
+	// selector.map_bits(|bit, index| if (bit == 1) { dna1 & (1 << index) } else { dna2 & (1 << index) })
+	// 注意 map_bits这个方法不存在。只要能达到同样效果，不局限算法
+	// 测试数据：dna1 = 0b11110000, dna2 = 0b11001100, selector = 0b10101010, 返回值 0b11100100
+	return dna1;
```

Comment:
This doesn't meet the requirement, check other people's work and see if you can understand.